### PR TITLE
raftstore: add tick registry

### PR DESCRIPTION
**Pull request by BusyJay**
## What have you changed? (mandatory)

Tick registry is used to avoid schedule a tick twice. It's very useful
when try to schedule tick lazily.

## What are the type of the changes? (mandatory)

- New feature

## How has this PR been tested? (mandatory)

unit tests

## Does this PR affect documentation (docs) or release note? (mandatory)

No.

## Does this PR affect tidb-ansible update? (mandatory)

No.

## Refer to a related PR or issue link (optional)

tikv/rfcs#23, tikv/tikv#4591

### CODE REVIEW

**Comment by Connor1996** [comment]
wrong RFC link

**Comment by BusyJay** [comment]
> wrong RFC link

Fixed.

File: src/raftstore/store/fsm/peer.rs [hunk 1]
```diff
@@ -582,10 +589,15 @@ impl<'a, T: Transport, C: PdClient> PeerFsmDelegate<'a, T, C> {
...
-    fn schedule_tick(&self, tick: PeerTick, timeout: Duration) {
+    fn schedule_tick(&mut self, tick: PeerTicks, timeout: Duration) {
+        if self.fsm.tick_registry.contains(tick) {
+            return;
+        }
         if is_zero_duration(&timeout) {
             return;
         }
+        trace!("schedule tick"; "tick" => ?tick, "timeout" => ?timeout, "region_id" => self.region_id(), "peer_id" => self.fsm.peer_id());
+        self.fsm.tick_registry.insert(tick);
```
Comment (Connor1996) [inline src/raftstore/store/fsm/peer.rs]: what if it failed to get mailbox or schedule peer tick, then the flag would be never removed

Comment (BusyJay) [reply src/raftstore/store/fsm/peer.rs]: The error can only happen when the node is shutting down. But it makes sense to keep registry clean.

File: src/raftstore/store/fsm/peer.rs [hunk 2]
```diff
@@ -65,6 +65,8 @@ pub struct DestroyPeerJob {
 
 pub struct PeerFsm {
     peer: Peer,
+    /// A registry for all scheduled ticks. This can avoid schedule ticks twice accidentally.
```
Comment (siddontang) [inline src/raftstore/store/fsm/peer.rs]: avoid scheduling ticks

File: src/raftstore/store/fsm/peer.rs [hunk 3]
```diff
@@ -608,7 +632,7 @@ impl<'a, T: Transport, C: PdClient> PeerFsmDelegate<'a, T, C> {
             .map(move |_| {
                 fail_point!(
                     "on_raft_log_gc_tick_1",
-                    peer_id == 1 && tick == PeerTick::RaftLogGc,
+                    peer_id == 1 && tick == PeerTicks::RAFT_LOG_GC,
                     |_| unreachable!()
                 );
                 if let Err(e) = mb.force_send(PeerMsg::Tick(tick)) {
```
Comment (Connor1996) [inline src/raftstore/store/fsm/peer.rs]: maybe add a comment to indicate that it's fine to not clean registry here cause the error happens when the node is shutting down

**Comment by hicqu** [comment]
LGTM

**Pull request merged**